server/gitrest/packages/gitrest-base/src/utils/redisFs/redisFsManager.ts

```diff
@@ -0,0 +1,457 @@
+	constructor(
+		redisParam: RedisParams,
+		redisOptions: IoRedis.RedisOptions,
+		redisFsConfig: RedisFsConfig,
+	) {
+		this.promises = new RedisFs(redisParam, redisOptions, redisFsConfig);
+	}
+}
+
+export class RedisFs implements IFileSystemPromises {
+	public readonly redisFs: Redis;
+	constructor(
+		redisParams: RedisParams,
+		redisOptions: IoRedis.RedisOptions,
+		private readonly redisFsConfig: RedisFsConfig,
+	) {
+		const redisClient = new IoRedis.default(redisOptions);
+		this.redisFs = new Redis(redisClient, redisParams);
+	}
+
+	/**
+	 * Asynchronously reads the entire contents of a file.
+	 * For more info on how isomorphic-git uses this, see:
+	 * https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L83
+	 */
+	public async readFile(
+		filepath: PathLike | FileHandle,
+		options?: { encoding?: null | undefined; flag?: OpenMode | undefined } | null,
+	): Promise<Buffer>;
+	public async readFile(
+		filepath: PathLike | FileHandle,
+		options: { encoding: BufferEncoding; flag?: OpenMode | undefined } | BufferEncoding,
+	): Promise<string>;
+	public async readFile(
+		path: PathLike | FileHandle,
+		options?: (ObjectEncodingOptions & { flag?: OpenMode | undefined }) | BufferEncoding | null,
+	): Promise<Buffer | string>;
+	public async readFile(
+		filepath: PathLike | FileHandle,
+		options?: any,
+	): Promise<Buffer | string> {
+		const filepathString = filepath.toString();
+		// Do not read packed-ref files which are not supported in r11s scenarios
+		if (filepathString.includes(packedRefsFileName)) {
+			return undefined;
+		}
+
+		const data = await executeRedisFsApi(
+			async () => this.redisFs.get<Buffer | string>(filepathString),
+			RedisFsApis.ReadFile,
+			RedisFSConstants.RedisFsApi,
+			this.redisFsConfig.enableRedisFsMetrics,
+			this.redisFsConfig.redisApiMetricsSamplingPeriod,
+			{
+				filepathString,
+			},
+			true,
+		);
+		// const data = await this.redisFs.get<Buffer | string>(filepathString);
+		return data;
+	}
+
+	/**
+	 * Asynchronously writes data to a file, replacing the file if it already exists.
+	 * For more info on how isomorphic-git uses this, see:
+	 * https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L101
+	 */
+	public async writeFile(
+		filepath: PathLike | FileHandle,
+		data:
+			| string
+			| NodeJS.ArrayBufferView
+			| Iterable<string | NodeJS.ArrayBufferView>
+			| AsyncIterable<string | NodeJS.ArrayBufferView>
+			| Stream,
+		options?:
+			| (ObjectEncodingOptions & {
+					mode?: Mode | undefined;
+					flag?: OpenMode | undefined;
+			  } & Abortable)
+			| BufferEncoding
+			| null,
+	): Promise<void> {
+		const filepathString = filepath.toString();
+		// Do not write packed-ref files which are not supported in r11s scenarios
+		if (filepathString.includes(packedRefsFileName)) {
+			return;
+		}
+
+		const result = await executeRedisFsApi(
+			async () => this.redisFs.set(filepath.toString(), data),
```

Comment:
The `redisFs` a.k.a. Redis client's `set()` method will call `JSON.stringify()` on this parameter. What kinds of values do we usually expect here? I don't recall for sure but I think this sometimes is not a string, but instead some binary data like Buffer/ArrayBuffer. Does `JSON.stringify()` work on those types?